### ISSUE TITLE
Installing Java Bindings

### ISSUE THREAD

**Issue by e1528532** [ElektraInitiative/libelektra#1083]
I wanted to try out the java bindings for my further work i inteded to do on the java bindings and related things. These are my experiences with the bindings, some things i was able to resolve, other things are still unclear. I will then try to update some of the documentations of jni and jna to help other users to get started easier.

# Getting Started

So first i had no real idea about how to install the java bindings. I quickly figured out that its probably a binding, and searched for bindings on the main page of libelektra. This lead me to the bindings folder which actually included the jna folder. On the tutorial main page the jna bindings are not listed though, i think we should add it there https://github.com/ElektraInitiative/libelektra/tree/master/doc/tutorials too.

So then i checked the documentation on https://github.com/ElektraInitiative/libelektra/tree/master/src/bindings/jna . Having to have jna.jar in the classpath was clear for me, but i wondered about the mentioned libelektra.jar . I wasn't really able to locate it anywhere in the system and that lead me to thinking that i have to compile the jni plugin which i thought will install this libelektra.jar in my system. I know that the jna documentation states

>or this directory which contains the elektra subdirectory

but i didn't figure out initially that this is the source of the libelektra.jar which i was searching for. 

# Stuggling with compiling the JNI plugin

## Setting the Java Path
So since i initially though that i need the jni plugin for the libelektra.jar i tried to compile elektra with this plugin. I headed over to the plugin documentation on https://github.com/ElektraInitiative/libelektra/tree/master/src/plugins/jni and then tried to compile it. I saw all these instructions with default-java and using ldconfig, but since i am using macOS Sierra these things are not available for me. So i decided to ignore that first and try to compile it right away using the stated command:

> cmake -DPLUGINS="...;jni" /path/to/libelektra

As this was the first plugin i tried to compile apart from just compiling elektra with the default plugins i was really confused about the correct cmake parameters (yes i haven't extensively read the documentation yet at that point...). So i actually ended up trying to compile elektra with 

>cmake -DPLUGINS="...;jni" ..

being in the build directory i created on the initial installation. Yes i actually kept the ... and then the plugin kept failing to compile. I always checked out the CMakeError.log file but it was totally non-obvious to figure out whats missing. Since the documentation was about changing some system paths before which i haven't done yet i figured that thats whats missing. But using macOS i had no idea where it searches for the files and what i have to change, until i saw in the CMakeError.log that it searches the java headers at the path 

> /Application/XCode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.12.sdk/System/Library/Frameworks/JavaVM.framework

Since the Current folder in that path is linked to the apple specific java implementation compilation does not work as the plugin expects java8. So after spending hours and hours i finally fixed that by changing the path link in the above mentioned location to point to the Oracle JDK folder, which is at /Library/Java/JavaVirtualMachines/jdk1.8.0_112.jdk . You also have to change the link of the Headers folder. I will update the documentation of the jni binding to explain these steps on a mac.

## It still doesn't work?

So after that it still didn't work. I slowly thought that the java plugin is maybe not working on OSX yet since its experimental. But after studying in the CMakeError.log for a last the i noticed that i actually had to specify the basic plugins which have to be included, like the resolver. After that i was finally able to compile the jni plugin successfully.

# Back to JNA
So then i tried again to make the jna bindings work. I tried again to search for the libelektra.jar which i though is installed along with the jni plugin now, but again failed to find it. So i compiled the source files anyway and it actually did work. Then i noticed that the libelektra.jar is actually just a jar of the elektra files in the bindings folder and if its not specified it just compiles those files directly. (Yes i should have probably tried that before spending hours on the jni plugin...)
Next it made me think if the JNI plugin was even necessary, so i compiled elektra again with the default settings and then the jna binding again and it still worked. So i that leads to the last thing

# JNI plugin vs JNA bindings

What is the actual difference now between the jni plugin and the jna bindings? JNI and JNA are to my understanding just both different ways of invoking native libraries in Java. But then why is the one a plugin and the other well - just a binding? And is my observation correct that the JNA bindings work independent of the JNI plugin? Its easy to figure out now how to use the JNA bindings in Java applications, but how would one use the JNI plugin? If my assumption that JNI and JNA do basically the same thing is correct, is it even necessary to have both implementations?

# TL;DR and TODOs

- [ ] Add JNA bindings to tutorial page
- [x] Explain in the jna tutorial what libelektra.jar is and where it comes from in more detail
- [x] Explain that the jni plugin is not required for the jna bindings
- [x] Update jni documentation on setting the java path on macOS and how to compile it there
- [x] Explain difference between JNI plugin and JNA bindings?

**Comment by markus2330** [comment]
Thanks for the comprehensive discussion of what you tried! Please keep up writing them. It makes it much clearer where problems in the documentation are! (And immediately help if someone else has the same question and is reading this)

> On the tutorial main page the jna bindings are not listed though

There currently is no tutorial covering java at the moment, thus it is not listed.

> which i was searching for.

Thanks, that clearly is a documentation issue.

> cmake -DPLUGINS="...;jni" ..

I avoided the ... in bcfc8d84276ca3a965540ccd6993e1055a5feade but Mac OS X is still not covered in the tutorial

> I will update the documentation of the jni binding to explain these steps on a mac.

Thanks!

> What is the actual difference now between the jni plugin and the jna bindings?

First: jni and jna are both java technologies that could be used to write bindings. jni is able to do more: It allows developers to integrate java in C Code.  But I think your actual question is what the difference between bindings and plugins is. I hope 2f210eeeeaacdd9e7069c3ab86c8f8d25e60ac1e makes it more clear?

> And is my observation correct that the JNA bindings work independent of the JNI plugin?

Yes, only plugins depend on bindings (at run-time), but not the other way round.

> but how would one use the JNI plugin?

Search for "mounting". And src/plugins/jni/README.md gives an example.

> If my assumption that JNI and JNA do basically the same thing is correct

No, bindings and plugins are completely different things.

**Comment by markus2330** [comment]
@Piankero if something of this text is not yet documented within the repo, can you please do that?

**Comment by Piankero** [comment]
This issue is unrelated to visible error messages.
As discussed in the Elektra meeting I may unassign myself.

**Comment by markus2330** [comment]
@Piankero this is a quite simple docu task. Didn't you install the Java bindings and can describe it for others?

**Comment by Piankero** [comment]
To be honest I am confused what `install` means in this context. I `used` the java binding for Lyrebird but that was such an obvious task that it is not worth a docu.

All I have done for my purposes was to copy the source of the binding to the new project (because it is not yet available on maven central) and renamed the pom.xml.in to pom.xml so mvn can load all dependencies (and I can possibly tweak the binding for additional print lines).

I do not (and actually think as of today its unnecessary) use command line compiling/installing as intellij does everything for me out of the box. I am very proficient with IJ and maven but when it comes to low level command line compiling I am actually the wrong person to document this.

**Comment by markus2330** [comment]
Yes, it would be enough what @MasterToney said today: to describe what the jna binding and jni plugin is about and which jars to copy and where to get them from.

maven central would be, of course, the best solution.
